### DISCUSSION TITLE
Version 4

### DISCUSSION THREAD

**Discussion by Gazook89** [naturalcrit/homebrewery#2879] [Ideas]
Opening this discussion to collect ideas in a 'next big version'.  There are other discussion threads that focus on more specific aspects of Homebrewery, which I'll link to, but for things that don't have or need their own thing, here we go:

a. Typescript
b. Responsive/Mobile UI (and codemirror 6?)
c. Two-way interaction between renderer and editor (d3.js?)

**Comment by 5e-Cleric** [comment]
AFAIK, CSS layers and brew themes are coming with v4, aren't they? 

I'll add to this the idea for the editor modes(pop out editor, or being able to change it to the right of the brew) as it is related to the responsive design. #2780 

And if we can, Less and Sass support would be nice to have too, as well as CSS Folding.  #2793

**Reply by calculuschild** [reply]
Themes are intended for v3. Just waiting for bits of cleanup and getting links to custom themes working.

**Comment by dbolack-ab** [comment]
> a. Typescript 

Like a full refactor?

> b. Responsive/Mobile UI (and codemirror 6?) 

Do people do this? try to author on phones or is there something I don't understand?

**Reply by calculuschild** [reply]
a) I'm actually pretty uneasy about Typescript. Every time i have used it, it adds so much bloat and boilerplate that any stability gained from typing is overwhelmed by illegibility. I would vote no on this.

b) Yeah... I am always surprised but people do try to edit on mobile with some frequency.

**Comment by ericscheid** [comment]
d) simultaneous editing by multiple authors

**Reply by 5e-Cleric** [reply]
This i've wanted for sooo long, i tend to do my best worldbuilding when i have someone to bounce crazy ideas off.

**Comment by 5e-Cleric** [comment]
e) spellcheck.
AFAIK, you can't use browser's spelling features in an element that is not editable(exc. using duplication of elements).
But we could add a js library for spellchecking
it could even be a toggle spellcheck, if it takes too much memory.

F) snippet preview?

~G) FontAwesome icon picker~ Done

~H) All selectable fonts as snippet to add~ Done

I) OUR OWN PDF RENDERING, i want bookmarks

~J) adding comments~ as if html comments didn't exist

K)BLEEDING

**Reply by calculuschild** [reply]
I think F, G, and H could be part of V3 if we want. H would actually be really easy and already kind of exists in the **Table -> Rune Table** snippet.

So yes I agree these would all be nice. But if they would work with V3 in a way that doesn't break anything, we don't have to wait to implement them. V4 would be more for changes that would break V3 (i.e., bleeding would require a lot of redesign of the CSS that would break existing documents, etc).

**Reply by dbolack-ab** [reply]
I have been beating my head on the PDF thing for a while. I've gotten close - but it honestly looks like it may require a proper markdown to PDF tool.

I'm still trying to get around that.

**Comment by ericscheid** [comment]
For adding comments to brews I simply use html comments.

```html
<!-- this is a comment -→
```

There's even a keyboard shortcut for it, and it is easy to select whole brew paragraphs (etc) and then comment them out.

(It does mean it still shows in the rendered html source. I'm OK with that.)

**Reply by 5e-Cleric** [reply]
you are completely right, i didn't even think about html comment, that syntax is terrible, we could add a nicer one

**Reply by dbolack-ab** [reply]
Most of the suggestions I see for the idea of comments are *terribly* wordy.  Assuming you want something smaller and wrappy...

Some options.

/* Comment */
// Comment //

Alternately, not wrappy
\comment
\c

**Reply by ericscheid** [reply]
A use case I have used is commenting out a segment of a paragraph.

> If you have (say) a table requiring a roll of 2d6 to determine creature of encounter and follow that with rolling d6 on another table for various variations `<!-- (e.g. is this a lone scout, or is this a hunting pack, etc) -->` … instead extend the table to accommodate rolling 2d12 for 2–24 lines, and fill in the extra lines with the variations. or roll 3d6 in first instance for 3–18 lines.

Inventing a new syntax could result in existing brews suddenly breaking (i.e. they've got plain text of `// Comment //`, but that now disappears on rendering because it is read as a comment). This qualifies as a breaking change, so not entirely off the table since we are talking v4. (It does mean converting legacy or v3 brews to v4 will be more complicated.)

**Reply by dbolack-ab** [reply]
The alternate is to suggest folks do something like `[Comment](#)` but that's... ugly.

An unwillingness to add new syntax for new features seems to be a common wall here.

**Reply by Gazook89** [reply]
Regarding "not wrappy":   I don't think that will work for the obvious reason that we don't know when to terminate the comment, unless it's only allowed on otherwise empty lines.

I do feel like it's likely best to stick with the existing syntax, which is HTML comments.  As with everything else regarding syntax it's better to stick with the *Standards* wherever possible, particularly for anything that isn't a new *thing*.  Yes, `<!-- comment -->` is very verbose...but it's well documented.  And, as noted, there is a shortcut for it (which is the same shortcut for CSS comments on the Style editor).  It already has syntax highlighting and because it is standard, all the editor themes take it into account.

HOWEVER

if we toss that all out the window, and we are thinking about v4 so breaking things is "fine", I like the semi-javascript approach of:

`Here is my // comment // that is inline`

and

```
Here is a

// comment

that is by itself on a line which is just a 'shortcut' because this also still works:

// comment //
```

Some other random ideas:
`/< comment >/`
`/' comment '/`
`<* comment *>`

**Reply by ericscheid** [reply]
> Here is my // comment // that is inline

In javascript, everything following the `//` is a comment. Not just the bit between the pair of `//`.

If you want an inline comment, it would be like this.
> Here is my /* comment */ that is inline

So .. that's *two* forms of markup for comments.

**Reply by Gazook89** [reply]
Yeah, that's why i said "semi-javascript".  I just like repeating a single character twice rather than complicating it.

However, I very truly do not have a horse in this race.  HTML comments seem fine to me.  They are visually distinctive from other text (which to me is a bonus for comments), and the others reasons i mentioned.

**Reply by 5e-Cleric** [reply]
i just hate the fact that to comment something i need 5 keystrokes  (`<` `Shift+1` `-` `-`) for

And before anyone says why don't you use the shortcut, it is not possible with my keyboard, i need shift for outputting `/`

**Reply by dbolack-ab** [reply]
It has been discovered recently that: 

marked provides `<? ?>` for comments
marked-smartypants provides `<!` which can run unenclosed until the end of a text block.

**Reply by 5e-Cleric** [reply]
must point out, some of these produce empty p elements with the comments inside, instead of just the comment.

**Comment by 5e-Cleric** [comment]
As talked about in Gitter today, we might want to add/change writing syntax for some things, like adding underline syntax:

`_this text would be underlined_` as syntax
and the shortcut could be `Ctrl + u`
in the html as `<u>this text would be underlined</u>`

also italics could use `<i>` if it doesnt already, and bold should probably use `<b>`, since we don't care for readers anyway.


- Highlight text syntax?

L. Copy button next to Code blocks

**Reply by dbolack-ab** [reply]
Italics and bold use \<em> and \<strong> respectively. My understanding ( which may be wrong or outdated ) is that emphasis doesn't have a *specific* requirement for how emphasis is applied - just that the standard practice is italics in many/most language setups. strong is similarly weakly defined. As such, the only thing making \*Thing\* Italicize **Thing** is the style sheet. :)

I am rather loosey-goosey on the whole notion of changing things in potentially brew-breaking ways but changing the meaning of \_ and \_\_ seems pretty reasonable. I'm not sure that many folks who do a lot fo markdown use it over \* patterns.  Is there any way to survey? Scan actual brews?

**Reply by 5e-Cleric** [reply]
The thing is that `em` and `strong` are usually better practice because you want screen readers to pick it up as strong or emphasized, but that is not an issue here, although there isnt much difference.

We tend not to want changing brew-breaking stuff, specially if there is no significant improval. But for v4 it'd be fine i reckon.

**Comment by 5e-Cleric** [comment]
Based on the discussion on Gitter chat, the next syntax ideas might be the ones actually used in v4:

|       ELEMENT        | v3 SYNTAX            | v4 SYNTAX            |
|:--------------------:|:---------------------|:---------------------|
|        headers       | `#` to `######` `header` | `#` to `######` `header` |
|      *italics*       | `*italics*`          |`*italics*`           |
|      **bold**        | `**bold**`           | `**bold**`           |
|  ***italicsbold***   | `***italicsbold*** ` | `***italicsbold*** ` |
| <ins>underline<ins/> | `<u>underline</u>`   | `_underline_`        |
|   ~strikethrough~    | `~strikethrough~` OR `~~strikethrough~~` | `~strikethrough~` OR `~~strikethrough~~` |
|     [link](href)     | `[link](href)`       | `[link](href)`       |
| <p align="right">right aligned text</p> | `{{text-align:right right aligned text}}` | `-: right aligned text` |
| <p align="left">left aligned text</p> | `{{text-align:left left aligned text}}` | `:- left aligned text` |
| <p align="center">center aligned text</p> | `{{text-align:center center aligned text}}` | `:-: center aligned text`|
| <ul><li>list element</li></ul> | `- list element` OR `* list element` OR `+ list element`|  `- list element`  (unsure if we should leave `*`) |
| <ol><li>ordered list element</li></ol> | `x. list element` OR `i list element`|`x. list element` OR `i list element`|
| :closed_book: | `:book:` | `:book:` |
| vertical <br><br> space | `::::` | `::::` |
| line <br>break | two spaces or `\` at end of sentence, or `<br>` | two spaces or `\` at end of sentence, or `<br>`|
| horizontal &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; space  | `horizontal :>>>>>>>> space`  | `horizontal :>>>>>>>> space` |
| <dl><dt>inline term</dt><dd>definition</dd></dl> | Inline term::definition | Inline term :: definition (enforce spaces around ::) |
---

### Not agreed upon yet
| ELEMENT | v3 SYNTAX | v4  PROPOSED SYNTAX |
|:-------:|:----------|:--------------------|
| - caption | {{text-align:right -- caption}} | ~caption |


In all of these cases the syntax can be somehow achieved using curly spans or block syntax, i don't want to add it every time or this table would be big as hell.

If i have left something out, be sure to comment so i can edit this table.

**Reply by Gazook89** [reply]
if wanting to avoid "doubling up" that regular markdown allows, you could add two more inline styles:

`~~double strikethrough~~` = `<del style='text-decoration: line-through double'>double strikethrough</del>`  (which doesn't show on Github...and you could use regular `span` tags too if you prefer.

and the other:

`__double underline__` = `<del style='text-decoration: underline double'>double underline</del>`

But maybe these "double" syntaxes are better reserved for some future syntax.  Meaning we purposefully *do not allow them* for now, even if commonmark allows them.

**Reply by dbolack-ab** [reply]
I did put in a PR for `__Foo__` to use the marked tag - aka strikethrough, completely forgetting about GFM being enabled.

**Reply by 5e-Cleric** [reply]
> But maybe these "double" syntaxes are better reserved for some future syntax. Meaning we purposefully do not allow them for now, even if commonmark allows them.

This seems like the most sensible perspective.

**Reply by T3sT3ro** [reply]
I wonder why it needs another syntax that is markdown-like, and why couldn't it use for example a standard that already exists AND is expressive and extandable, like for example [markdoc](https://markdoc.dev/) (alternatively AsciiDoc, although markdoc looks a lot more intuitive)?

What I am mainly focused on is their standard tag system would unify a bunch of different syntax extensions that are (as far as I'm aware) present currently in the syntax. There could also be several other areas of improvement regarding that: existing extensions (for example for VSCode and other allowing for working with mdoc syntax), offloading a bunch of development efforts to existing markdoc packages for handling content generation and implementing custom rules in directives in terms of tags and new markdoc nodes. Furthermore a frontmatter for defining global settings, which are colocated with the content.

**Reply by dbolack-ab** [reply]
Overlooked this earlier.

This is a pretty heavy lift. @calculuschild Am I mistaken?

**Reply by calculuschild** [reply]
@dbolack-ab Are you referring to the markdoc thing?

My understanding is it is mainly just an alternative to `marked.js` which we currently use, with the main feature being that it includes a `{$ tag $}` system which is essentially what our own `{{ }}` syntax does.

- `Marked.js` already supports extensions for custom syntax.
- `Marked.js` already offloads much development by nature of its existing packages we can choose from.
- We already have a `{{tag}}` syntax.

So at first pass, I see no compelling reason to switch tools when our existing solution already provides all the same features. Would definitely be a heavy lift in any case.

**Reply by dbolack-ab** [reply]
> @dbolack-ab Are you referring to the markdoc thing?


I was.

**Reply by dbolack-ab** [reply]
Created #3908 for the paragraph alignment trio.

**Reply by dbolack-ab** [reply]
Regarding the `*` mode for ULs. I personally come to that pattern from word-processing which is probably where this comes from in the spec. I can't remember the last one I used that didn't convert some version of `/^*/` into a bulleted list.

I expect removing that would cause more confusion than removing the other two modes would ( not that I have strong feelings about doing that instead )

**Reply by dbolack-ab** [reply]
Added #3909 For evaluating the non-breaking space token.

**Reply by 5e-Cleric** [reply]
> Regarding the `*` mode for ULs. I personally come to that pattern from word-processing which is probably where this comes from in the spec. I can't remember the last one I used that didn't convert some version of `/^*/` into a bulleted list.
> 
> I expect removing that would cause more confusion than removing the other two modes would ( not that I have strong feelings about doing that instead )

I don't care the one we change, but `-` is the only one in my keyboard i can use without additional keys.

**Reply by dbolack-ab** [reply]
> I don't care the one we change, but `-` is the only one in my keyboard i can use without additional keys.

Unless we wish to reuse one or more of those tokens elsewhere I don't think this is a valuable change.

**Reply by Gazook89** [reply]
I think we keep all three syntaxes for the lists.  But I also see that i already said that in the next comment thread below this one, where Eric explains why.

**Reply by 5e-Cleric** [reply]
i say again, i see why anyone would want 2, just don't see where the third one comes in handy. but alright, i'm not going to force this change.

**Comment by ericscheid** [comment]
Using `-` and `+` as alternative bullet list forms is actually useful .. because they get rendered as separate `<ul>`.

This brew source..
```
+ item
+ item
+ item

- item
- item
```

.. renders to
```html
<ul>
  <li>item</item>
  <li>item</item>
  <li>item</item>
</ul>

<ul>
  <li>item</item>
  <li>item</item>
</ul
```

This is useful. I have used this when writing up long bullet list outlines where I want to keep different sections separate.

**Reply by 5e-Cleric** [reply]
Is there a real difference(in brew) apart form visual separation, which can also be achieved using a colon?

**Reply by ericscheid** [reply]
Mostly not. 

Unless I'm using CSS like `.page#p2 ul:nth-of-kind(3)` .. because it's now all one huge bullet list with some `<div class=blank>` inserted inside list items.

<img width="338" alt="image" src="https://github.com/naturalcrit/homebrewery/assets/2605467/efc4e402-3ef8-419f-a4f0-5363b52ec9a9">

I would have to resort to hacking up my source with `{{curly-blocks}}` to get the separation. A step backwards.

**Reply by Gazook89** [reply]
I am doubtful of how often this functionality gets used, but even so I think I am with Eric on this one. To match the current output we'd have to create more CSS and otherwise twist around to get the desired effect...and any effort spent doing that isn't worth breaking from the commonmark spec.  And, the list syntax isn't `+item`, it's with the space `+ item`....so I'm not sure that we gain anything from dropping those other list indicators anyway.

**Reply by dbolack-ab** [reply]
That is unfortunate.

**Reply by 5e-Cleric** [reply]
i still don't think we need three separate syntaxes for the same thing.

**Reply by dbolack-ab** [reply]
> i still don't think we need three separate syntaxes for the same thing.

Need? No. But lacking a reason to remove them, we maybe shouldn't?

**Comment by dbolack-ab** [comment]
What about adding thumbnail generation? It is fairly easy.

**Reply by 5e-Cleric** [reply]
you mean outputting the first page of a brew as a png?

**Reply by dbolack-ab** [reply]
Essentially

**Comment by calculuschild** [comment]
A couple more possibilities:

##### Index Markers (already in #3133)

Basically `#label` (no space after `#`) to indicate an entry that should appear in an index, plus some sugar added on to indicate synonyms/parent/child relationships between entries. `#index | synonym` or similar if I recall.

Which makes me think perhaps we can extend headers in a similar way to guide TOC generation instead of relying upon CSS `--excludeTOC` tags? Given that indices and table of contents fall into a similar class of "auto-generated structure".

##### TOC Markers

Some permutation of `# Header @` where `@` is some symbol or string we devise to indicate that a header should be excluded/included from the Table of Contents.

**Reply by dbolack-ab** [reply]
> ##### TOC Markers
> Some permutation of `# Header @` where `@` is some symbol or string we devise to indicate that a header should be excluded/included from the Table of Contents.

Currently, the ToC generator runs entirely via DOM scraping, so anything like this would need to override the header parsing and add this field as a HTML attribute on the `<H#>` in question. I don't think this is a bad idea, but I'm unsure if this might interfere with the HeaderID addon?

**Reply by 5e-Cleric** [reply]
I am not sure we want to pursue this route at all. I think, that indexes should be done by hand, once the brew is done. Same for ToC, is it a lot of work? absolutely. Could we help users with it? sure. But no feature is going to be perfect for any user. I think we should choose whatever method based on:

- ease of use
- ease to code
- flexibility of the method to create different outputs.

Perhaps the best thing we could do is some widget to let the user look for instances of a word in the text, like the search text feature, then collect the instances. Not while writing and cluttering the syntax, but later looking for that text programatically. And let me clarify that, ONLY when generating, not constantly.

@dbolack-ab @calculuschild What do you think?

**Reply by dbolack-ab** [reply]
> I am not sure we want to pursue this route at all. I think, that indexes should be done by hand, once the brew is done. Same for ToC, is it a lot of work? absolutely. Could we help users with it? sure. But no feature is going to be perfect for any user. I think we should choose whatever method based on:

I'm at a loss here, particularly when the code has been done for months....

**Reply by 5e-Cleric** [reply]
I do see how my comment would sound like nonsense

**Reply by dbolack-ab** [reply]
> I do see how my comment would sound like nonsense

No, It doesn't come across as nonsense. However, it is antithetical to the idea of the Rich Indexes code.

That said a "simple" index that operated in this way might also be useful. If I was to select one or the other, I would, as a user, find the simple option very disappointing on a product that *needed* an index..

**Comment by Gazook89** [comment]
Table column width syntax as talked about here #1528

**Reply by 5e-Cleric** [reply]
Do we need to wait till v4 for that? as long as we assume no numbers as the current behaviour we should be able to do this without breaking, no?

**Reply by dbolack-ab** [reply]
Working version: https://github.com/calculuschild/marked-extended-tables/pull/456

**Comment by calculuschild** [comment]
Another proposal for V4: Change `::::` "vertical spacing" to `\\\\`, and have it emit `<br>` instead of `<div class="blank">`

Already in CommonMark, we have a way to generate `<br>` by ending a line with two spaces or a `\`. So you can do something like this:

```
text\
\
\
\
more text
```

To get a series of breaks. It seems we could just build on that to allow `\\\` for multiple breaks:

```
text
\\\
more text
```

In the same way we already use `:::`. In other words, we extend the "line break" syntax and potentially remove conflicts with other `:`-based syntax. Thoughts?

**Reply by dbolack-ab** [reply]
> Another proposal for V4: Change `::::` "vertical spacing" to `\\\\`, and have it emit `<br>` instead of `<div class="blank">`

I just tested this and it doesn't look like it's directly supported by marked yet.

```
> text\
```
The rest was an easy switch. 

If you were expecting our plugin to add it, is the  above scenario supposed to emit 

```
text<br>
```

or 

```
text
```

followed by the parent container closing?

**Reply by ericscheid** [reply]
> Already in CommonMark, we have a way to generate <br> by ending a line with two spaces or a \. So you can do something like this:
> To get a series of breaks. It seems we could just build on that to allow `\\\\\` for multiple breaks:

Sounds elegant .. but with one wrinkle in the ointment. 

The `\` in commonmark is used to escape the following character. That is, ignore any syntactical meaning, applying the literal meaning — in this case ignoring the usually ignored inner-paragraph line break in the source text, and substituting in an actual line-break in the render text.

Thus, a question: does three backslashes at the end of a line (`\\\¶`) in source-text represent syntax for three line breaks in the render-text, or represent an actual literal backslash character ("\")  followed by a linebreak in the render-text?

**Reply by 5e-Cleric** [reply]
I prefer colons for this, its way easier to input in my keyboard, because . and shift are so close together, as opposed to the backslash where i have to do 7 + shift

**Comment by Gazook89** [comment]
# Table of Contents

I think we can move away from using headers in the formatting of the Table of Contents.  So instead of this:

```md
{{toc,wide
# Contents

- ### [{{ cat}}{{ 3}}](#p3)
  - #### [{{ dog}}{{ 3}}](#p3)
}}
```
we have this:
```md
{{toc,wide
# Contents

- [{{ cat}}{{ 3}}](#p3)
  - [{{ dog}}{{ 3}}](#p3)
}}
```
Semantically, a table of contents shouldn't be written as headers itself.  Further, it would make the HeaderNav easier, because right now we have special logic to keep the ToC headers from appearing in the HeaderNav.  

*Edit*

We could go further in simplifying the ToC if we *didn't care* about semantics.....Instead of `span`s, we could use an html node to separate the **link text** from the **link page number** that has a simpler markdown syntax than `{{ text}}`.  Something like `*italic*` syntax.  This would give us an HTML node that we can apply CSS too but be simpler to read/write:

```md
{{toc,wide
# Contents

- [*cat* *3*](#p3)
  - [*dog*  *3*](#p3)
}}
```

**Reply by dbolack-ab** [reply]
I don't understand the assertion it is semantically incorrect.

**Reply by 5e-Cleric** [reply]
> The [i](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-i-element) element [represents](https://html.spec.whatwg.org/multipage/dom.html#represents) a span of text in an alternate voice or mood, or otherwise offset from the normal prose in a manner indicating a different quality of text, such as a taxonomic designation, a technical term, an idiomatic phrase from another language, transliteration, a thought, or a ship name in Western texts.

By this definition, the `i` element is not ideal in this use.

If we are interested in keeping the semantics in what will be output as a pdf, we should be using the following structure:

```html
<nav>
  <h2>Table of Contents</h2>
  <ol>
    <li><a href="#chapter-1">Chapter 1</a>
    <li><a href="#chapter-2">Chapter 2</a>
    <li><a href="#chapter-3">Chapter 3</a>
  </ol>
</nav>
```

Which in an only Markdown approximation would be:

```md
{{toc,wide
# Contents

1. [{{cat}} {{3}}](#p3)
  1.1. [{{dog}} {{3}}](#p3)
}}
```
I say, we keep using spans and `ul` for this. So i agree with just removing the headers from the syntax.

**Reply by Gazook89** [reply]
> We could go further in simplifying the ToC if we **didn't care** about semantics

Yeah I know italic element isn’t appropriate, hence the caveat above. 

But I raised the point because using a simpler markdown syntax, in this case stealing the `i`/`*…*` would provide much more visual clarity to the toc than our current approach jam packed with 4 brackets and two inscrutable required ` `/spaces inside those brackets.  And again, visually, they’d appear as just non italic text. 

And it doesn’t have to be the italics syntax that is borrowed. It could be another simple syntax, like inline code backticks.

**Reply by Gazook89** [reply]
And I know that in one breath I'm saying "lets remove the header syntax because it's not correct" and in another i'm saying "if didn't care about being correct, lets use some wrong syntax".  

I don't think the header syntax is correct *and* it adds a lot of extra characters and visual clutter.

I think replacing the span syntax `{{ text}}{{ page}}` with `*text* *page*`is possibly worth the incorrect semantics for the ease of use.

**Reply by dbolack-ab** [reply]
> > The [i](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-i-element) element [represents](https://html.spec.whatwg.org/multipage/dom.html#represents) a span of text in an alternate voice or mood, or otherwise offset from the normal prose in a manner indicating a different quality of text, such as a taxonomic designation, a technical term, an idiomatic phrase from another language, transliteration, a thought, or a ship name in Western texts.

It is also a common practice for parts of ToCs, Appendices, and indices. Though to to be properly pedant here, we should be using `<mark>` , a class, or a style injection rather than using a tag that does the EXACT same thing. 

> I say, we keep using spans and `ul` for this. So i agree with just removing the headers from the syntax.

It's dumb redundancy, but sure.

**Reply by calculuschild** [reply]
The main reason for the two spans is to finagle the dot leaders working correctly. If you can figure out the CSS to remove one or both of those spans and still get the styling we want I am OK with that.

But I agree we shouldn't just substitute italics.

**Reply by Gazook89** [reply]
Quote from 5e's comment:

```
{{toc,wide
# Contents

1. [{{cat}} {{3}}](#p3)
  1.1. [{{dog}} {{3}}](#p3)
}}
```

Just noting that the above is *not* a functioning ToC, because it is missing the ` `/`space` before `cat`, `dog`, `3`, and `3`.  As-is, those would be empty spans with each of those strings as class names.  

And I'm noting it because it supports the point that the `{{ }}` span syntax is not intuitive when you want a `span` with no additional attributes, and that's pretty much every span in a ToC.  

Not a big deal when the ToC is auto-generated, just when someone is typing it out.

-----

That said, i'm happy with just removing the header syntax but otherwise leaving it alone.

**Reply by ericscheid** [reply]
> And I'm noting it because it supports the point that the {{ }} span syntax is not intuitive when you want a span with no additional attributes, and that's pretty much every span in a ToC.

We could write it as this, instead:
```
1. [{{entry: cat}} {{page: 3}}](#p3)
  1.1. [{{entry: dog}} {{page: 3}}](#p3)
```
I don't much like this approach (putting signs on things to explain how to use is a clear UX smell).

Alternatively, we could change the approach to (post-render) tweak the html links inside a `{{toc}}` block, and thus start with much simpler markdown of:
```
{{toc
1. [toc entry text](#p5 "106")
}}
```
Here the link title attribute is plucked out by a post-render plug-in to be the displayed page number.

**Reply by 5e-Cleric** [reply]
hmmm

**Comment by Gazook89** [comment]
### Font Imports

It'd be nice to make font imports easier.  So that in the Style Editor, one could do this:

```
/* [@font "Inter" variable] */

// or

/* [@font "Inter" static 400 100 700 italic bold normal] */
```

Or similar, and essentially get this:

```
@font-face {
  font-family: 'Inter Variable';
  font-style: normal;
  font-display: swap;
  font-weight: 100 900;
  src: url(https://cdn.jsdelivr.net/fontsource/fonts/inter:vf@latest/latin-wght-normal.woff2) format('woff2-variations');
  unicode-range: U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+0304,U+0308,U+0329,U+2000-206F,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD;
}
```

This utilizes the CDN for fontsource.org, which is basically Google Fonts (same library, plus some more).  It's *not* google fonts, because Google changes their fonts with updates and are not versioned, whereas fontsource can be (or like above, set with `latest`).  And it doesn't have Google tracking.  

Fontsource offers two regular patterns for their URLs, depending on variable or static:

```
// static
https://cdn.jsdelivr.net/fontsource/fonts/{id}@{version}/{subset}-{weight}-{style}.{extension}
 
// variable
`https://cdn.jsdelivr.net/fontsource/fonts/{id}:vf@{version}/{subset}-{axes}-{style}.woff2
```

I would expect this to be something entered in the Style Editor, not the Markdown/Brew editor, so I think it'd be the first foray into processing a new syntax in the CSS document?  And to keep it valid CSS, it would be wrapped in a CSS comment syntax, in case someone is taking the CSS portion and trying to use it elsewhere.

**Reply by 5e-Cleric** [reply]
I don't see that as being easier, if anything, abstracting it makes it harder to understand what each value is.
using font face you can see the numbers are font weights, not to mention you would still need the url there.
Opposed to this idea in this stage.

**Comment by Gazook89** [comment]
We should change the `.footnote` class to `.footer`.  We are using them to show the current section alongside the page number snippet, so it's more a footer than a footnote.  In #2885 we talk about adding a `.noFooter` class to the TOC snippet, and so if we do that we should make sure it's consistent.

**Comment by Gazook89** [comment]
Change the automatically created `p` tags in the `columnWrapper` to have some associated class name.  These tags:

```html
<div class="columnWrapper">
  <p>&nbsp;</p>       <!---- here ----> 
  <div class="columnSplit"></div>
  <p>&nbsp;</p>       <!---- here ----> 
</div>
```
Something to distinguish them from other p tags, enabling the use of `:not(.hb-column-hack)` or something to avoid selecting them in CSS.

**Reply by 5e-Cleric** [reply]
i'm still not sure why they exist in the first place.

**Reply by dbolack-ab** [reply]
They don't exist consistantly, either

**Comment by 5e-Cleric** [comment]
I am condensing this information and reopening another dicussion or issue, as this one is basically outdated and cluttered.
